File: Sources/XMTP/Messages/SealedInvitationV1.swift
//
//  SealedInvitationV1.swift
//
//
//  Created by Pat Nakajima on 11/26/22.
//

import Foundation
import XMTPProto

typealias SealedInvitationV1 = Xmtp_MessageContents_SealedInvitationV1

extension SealedInvitationV1 {
	init(headerBytes: Data, ciphtertext: CipherText, header _: SealedInvitationHeaderV1? = nil) {
		self.init()
		self.headerBytes = headerBytes
		ciphertext = ciphtertext
	}

	var header: SealedInvitationHeaderV1 {
		do {
			return try SealedInvitationHeaderV1(serializedData: headerBytes)
		} catch {
			return SealedInvitationHeaderV1()
		}
	}

	func getInvitation(viewer: PrivateKeyBundleV2) throws -> InvitationV1 {
		let header = header

		var secret: Data

		if !header.sender.identityKey.hasSignature {
			throw SealedInvitationError.noSignature
		}

		if viewer.identityKey.matches(header.sender.identityKey) {
			secret = try viewer.sharedSecret(peer: header.recipient, myPreKey: header.sender.preKey, isRecipient: false)
		} else {
			secret = try viewer.sharedSecret(peer: header.sender, myPreKey: header.recipient.preKey, isRecipient: true)
		}

		let decryptedBytes = try Crypto.decrypt(secret, ciphertext, additionalData: headerBytes)
		let invitation = try InvitationV1(serializedData: decryptedBytes)

		return invitation
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/SealedInvitationV1.kt
package org.xmtp.android.library.messages

import com.google.protobuf.kotlin.toByteString
import org.xmtp.android.library.CipherText
import org.xmtp.android.library.Crypto
import org.xmtp.android.library.XMTPException

typealias SealedInvitationV1 = org.xmtp.proto.message.contents.Invitation.SealedInvitationV1

class SealedInvitationV1Builder {
    companion object {
        fun buildFromHeader(headerBytes: ByteArray, ciphtertext: CipherText): SealedInvitationV1 {
            return SealedInvitationV1.newBuilder().also {
                it.headerBytes = headerBytes.toByteString()
                it.ciphertext = ciphtertext
            }.build()
        }
    }
}

val SealedInvitationV1.header: SealedInvitationHeaderV1
    get() = SealedInvitationHeaderV1.parseFrom(headerBytes)

fun SealedInvitationV1.getInvitation(viewer: PrivateKeyBundleV2?): InvitationV1 {
    val header = header
    if (!header.sender.identityKey.hasSignature()) {
        throw XMTPException("No signature")
    }
    val secret = if (viewer != null && viewer.identityKey.matches(header.sender.identityKey)) {
        viewer.sharedSecret(
            peer = header.recipient,
            myPreKey = header.sender.preKey,
            isRecipient = false
        )
    } else {
        viewer?.sharedSecret(
            peer = header.sender,
            myPreKey = header.recipient.preKey,
            isRecipient = true
        ) ?: byteArrayOf()
    }
    val decryptedBytes =
        Crypto.decrypt(secret, ciphertext, additionalData = headerBytes.toByteArray())
    return InvitationV1.parseFrom(decryptedBytes)
}
